File: Problems/Algorithms/377. Combination Sum IV/CombinationSumIV.kt
// The following solution leads to TLE
class Solution {
    private val curr: MutableList<Int>
    private var res: Int
    
    init {
        curr = mutableListOf()
        res = 0
    }
    
    fun combinationSum4(nums: IntArray, target: Int): Int {
        aux(nums, target, 0)
        return res
    }
    
    private fun aux(nums: IntArray, target: Int, start: Int): Unit {
        if (target == 0) {
            res++
            return
        }
        
        if (target > 0) {
            for (i in start..nums.size-1) {
                curr.add(nums[i])
                aux(nums, target-nums[i], start)
                curr.removeAt(curr.size - 1)
            }
        }
    }
}

File: Problems/Algorithms/377. Combination Sum IV/combination_sum_IV.py
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = []

        while len(dp) < target:
            dp.append(0)

            for num in nums:
                curr = len(dp)
                if num == curr:
                    dp[-1] += 1
                elif num < curr:
                    dp[-1] += dp[curr-num-1]

        return dp[-1]

